lib/measure-text.js

```diff
@@ -1,31 +1,114 @@
 'use strict';
 
-const path = require('path');
 const PDFDocument = require('pdfkit');
-let doc = new PDFDocument({size: 'A4', layout: 'landscape'});
-
-// Attempt to use a particular font.
-// callback: (optional) takes an error if it failed.
-function loadFont(path, callback) {
-  try {
-    doc = doc.font(path);
-    if (callback) { callback(null); }
-  } catch(err) {
-    doc = doc.font('Helvetica-Bold');
-    if (callback) { callback(err); }
+
+class PDFKitTextMeasurer {
+  constructor() {
+    this.document = new PDFDocument({ size: 'A4', layout: 'landscape' })
+      .fontSize(11);
+  }
+
+  useFont(fontPath) {
+    return new Promise((resolve, reject) => {
+      try {
+        this.document.font(fontPath);
+      } catch (e) {
+        reject(e);
+        return;
+      }
+      resolve(this);
+    });
+  }
+
+  widthOf(str) {
+    return this.document.widthOfString(str);
+  }
+
+  static create(fontPath, fallbackFontPath) {
+    return new this().useFont(fontPath)
+      .catch(err => {
+        if (fallbackFontPath) {
+          console.log(`Unable to load font at ${fontPath}. Trying fallback font ${fallbackFontPath}.`);
+          return new this().useFont(fallbackFontPath);
+        } else {
+          throw err;
+        }
+      });
   }
 }
 
-loadFont(path.join(__dirname, '..', 'Verdana.ttf'), function (err) {
-  if (err && process.env.FALLBACK_FONT_PATH) {
-    loadFont(process.env.FALLBACK_FONT_PATH);
+class QuickTextMeasurer {
+  constructor(baseMeasurer) {
+    Object.assign(this, { baseMeasurer });
```

Comment:
Injecting it makes testing a little easier, because I can place the spy on the measurer instance. Caching code is notoriously difficult to test so I feel it's important to test that the cache object doesn't call through to the base object when it's not being used. I'll take a look though; maybe that is easily changed now. When this is all synchronous it'll be way simpler.

Reply:
This is working just fine by mocking the method on the prototype instead!